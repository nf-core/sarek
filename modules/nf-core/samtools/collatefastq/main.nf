process SAMTOOLS_COLLATEFASTQ {
    tag "$meta.id"
    label 'process_low'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/samtools:1.21--h50ea8bc_0' :
        'biocontainers/samtools:1.21--h50ea8bc_0' }"

    input:
    tuple val(meta), path(input)
    tuple val(meta2), path(fasta)
    val(interleave)

    output:
    tuple val(meta), path("*_{1,2}.fq.gz")     , optional:true, emit: fastq
    tuple val(meta), path("*_interleaved.fq")  , optional:true, emit: fastq_interleaved
    tuple val(meta), path("*_other.fq.gz")     , emit: fastq_other
    tuple val(meta), path("*_singleton.fq.gz") , optional:true, emit: fastq_singleton
    path "versions.yml"                        , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def args2 = task.ext.args2 ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    def reference = fasta ? "--reference ${fasta}" : ""
    def output =    (interleave && ! meta.single_end) ? "> ${prefix}_interleaved.fq"                     :
                    meta.single_end                   ? "-1 ${prefix}_1.fq.gz -s ${prefix}_singleton.fq.gz" :
                    "-1 ${prefix}_1.fq.gz -2 ${prefix}_2.fq.gz -s ${prefix}_singleton.fq.gz"

    """
    samtools collate \\
        $args \\
        --threads $task.cpus \\
        ${reference} \\
        -O \\
        $input \\
        . |

    samtools fastq \\
        $args2 \\
        --threads $task.cpus \\
        ${reference} \\
        -0 ${prefix}_other.fq.gz \\
        $output

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        samtools: \$(echo \$(samtools --version 2>&1) | sed 's/^.*samtools //; s/Using.*\$//')
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    def empty = "echo '' | gzip "
    def singletoncommand = "${empty}> ${prefix}_singleton.fq.gz"
    def interleavecommand = interleave && !meta.single_end ? "${empty}> ${prefix}_interleaved.fq.gz" : ""
    def output1command = !interleave ? "${empty}> ${prefix}_1.fq.gz" : ""
    def output2command = !interleave && !meta.single_end ? "${empty}> ${prefix}_2.fq.gz" : ""

    """
    ${output1command}
    ${output2command}
    ${interleavecommand}
    ${singletoncommand}
    ${empty}> ${prefix}_other.fq.gz

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        samtools: \$(echo \$(samtools --version 2>&1) | sed 's/^.*samtools //; s/Using.*\$//')
    END_VERSIONS
    """
}
