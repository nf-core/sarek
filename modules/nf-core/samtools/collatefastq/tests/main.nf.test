nextflow_process {

    name "Test Process SAMTOOLS_COLLATEFASTQ"
    script "../main.nf"
    process "SAMTOOLS_COLLATEFASTQ"

    tag "modules"
    tag "modules_nfcore"
    tag "samtools"
    tag "samtools/collatefastq"

    test("human - bam - paired_end") {

        when {
            process {
                """
                input[0] = [
                         [ id:'test', single_end:false ], // meta map
                         file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/bam/test.paired_end.sorted.bam', checkIfExists: true)
                     ]
                input[1] = [
                         [ id:'test' ], // meta map
                         file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true)
                     ]
                input[2] = false
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out.fastq,
                                  process.out.versions).match() }
            )
        }

    }

    test("human - bam - single_end") {

        when {
            process {
                """
                input[0] = [
                         [ id:'test', single_end:true ], // meta map
                         file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.single_end.bam', checkIfExists: true)
                     ]
                input[1] = [
                         [ id:'test' ], // meta map
                         file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true)
                     ]
                input[2] = false
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out.fastq_other,
                                  process.out.versions).match() }
            )
        }

    }

    test("human - cram") {

        when {
            process {
                """
                input[0] = [
                         [ id:'test', single_end:false ], // meta map
                         file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/cram/test.paired_end.sorted.cram', checkIfExists: true)
                     ]
                input[1] = [
                         [ id:'test' ], // meta map
                         file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true)
                     ]
                input[2] = false
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out.fastq,
                                  process.out.versions).match() }
            )
        }

    }

    test("human - bam - paired_end - interleaved") {

        when {
            process {
                """
                input[0] = [
                         [ id:'test', single_end:false ], // meta map
                         file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/bam/test.paired_end.sorted.bam', checkIfExists: true)
                     ]
                input[1] = [
                         [ id:'test' ], // meta map
                         file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true)
                     ]
                input[2] = true
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out.fastq_interleaved,
                                  process.out.fastq_singleton,
                                  process.out.fastq,
                                  process.out.versions).match() }
            )
        }

    }

    test("human - bam - paired_end -stub") {
        options "-stub"
        when {
            process {
                """
                input[0] = [
                         [ id:'test', single_end:false ], // meta map
                         file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/bam/test.paired_end.sorted.bam', checkIfExists: true)
                     ]
                input[1] = [
                         [ id:'test' ], // meta map
                         file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true)
                     ]
                input[2] = false
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out.fastq,
                                  process.out.versions).match() }
            )
        }

    }

    test("human - bam - single_end - stub") {
        options "-stub"
        when {
            process {
                """
                input[0] = [
                         [ id:'test', single_end:true ], // meta map
                         file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.single_end.bam', checkIfExists: true)
                     ]
                input[1] = [
                         [ id:'test' ], // meta map
                         file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true)
                     ]
                input[2] = false
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out.fastq_other,
                                  process.out.versions).match() }
            )
        }

    }

    test("human - cram - stub") {
        options "-stub"
        when {
            process {
                """
                input[0] = [
                         [ id:'test', single_end:false ], // meta map
                         file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/cram/test.paired_end.sorted.cram', checkIfExists: true)
                     ]
                input[1] = [
                         [ id:'test' ], // meta map
                         file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true)
                     ]
                input[2] = false
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out.fastq,
                                  process.out.versions).match() }
            )
        }

    }

    test("human - bam - paired_end - interleaved - stub") {
        options "-stub"
        when {
            process {
                """
                input[0] = [
                         [ id:'test', single_end:false ], // meta map
                         file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/bam/test.paired_end.sorted.bam', checkIfExists: true)
                     ]
                input[1] = [
                         [ id:'test' ], // meta map
                         file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true)
                     ]
                input[2] = true
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out.fastq_interleaved,
                                  process.out.fastq_singleton,
                                  process.out.fastq,
                                  process.out.versions).match() }
            )
        }

    }



}
