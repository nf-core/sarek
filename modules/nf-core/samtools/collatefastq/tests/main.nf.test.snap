{
    "human - bam - paired_end - interleaved - stub": {
        "content": [
            [
                
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test_singleton.fq.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                ]
            ],
            [
                
            ],
            [
                "versions.yml:md5,6359bf4b480f99e14e14b9e7dcb2fd5a"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-16T07:50:35.483719031"
    },
    "human - bam - single_end": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": true
                    },
                    "test_other.fq.gz:md5,a6c101a06b5c9d5f8b91c0acd4ac5045"
                ]
            ],
            [
                "versions.yml:md5,6359bf4b480f99e14e14b9e7dcb2fd5a"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-16T07:49:31.667042764"
    },
    "human - bam - paired_end -stub": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    [
                        "test_1.fq.gz:md5,68b329da9893e34099c7d8ad5cb9c940",
                        "test_2.fq.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ]
            ],
            [
                "versions.yml:md5,6359bf4b480f99e14e14b9e7dcb2fd5a"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-16T07:50:04.607920717"
    },
    "human - cram - stub": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    [
                        "test_1.fq.gz:md5,68b329da9893e34099c7d8ad5cb9c940",
                        "test_2.fq.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ]
            ],
            [
                "versions.yml:md5,6359bf4b480f99e14e14b9e7dcb2fd5a"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-16T07:50:25.354607106"
    },
    "human - cram": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    [
                        "test_1.fq.gz:md5,1cf671980643af6c1148ae5e8e94e350",
                        "test_2.fq.gz:md5,38c1e9829115f9025f95435c5a4373d3"
                    ]
                ]
            ],
            [
                "versions.yml:md5,6359bf4b480f99e14e14b9e7dcb2fd5a"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-16T07:49:43.12646213"
    },
    "human - bam - paired_end - interleaved": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test_interleaved.fq:md5,4f2b93d492f0442fa89b02532c9b3530"
                ]
            ],
            [
                
            ],
            [
                
            ],
            [
                "versions.yml:md5,6359bf4b480f99e14e14b9e7dcb2fd5a"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-16T07:49:53.799565712"
    },
    "human - bam - single_end - stub": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": true
                    },
                    "test_other.fq.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                ]
            ],
            [
                "versions.yml:md5,6359bf4b480f99e14e14b9e7dcb2fd5a"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-16T07:50:15.487972669"
    },
    "human - bam - paired_end": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    [
                        "test_1.fq.gz:md5,1cf671980643af6c1148ae5e8e94e350",
                        "test_2.fq.gz:md5,38c1e9829115f9025f95435c5a4373d3"
                    ]
                ]
            ],
            [
                "versions.yml:md5,6359bf4b480f99e14e14b9e7dcb2fd5a"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-16T07:49:19.805942524"
    }
}